src/entities/road.lua

```diff
@@ -25,19 +52,84 @@ function road:draw()
                        segment.body:getAngle())
   end
   love.graphics.polygon("fill",
-    self.frontier.body:getWorldPoints(self.frontier.shape:getPoints())
+    self.frontier.main.body:getWorldPoints(self.frontier.main.shape:getPoints())
+  )
+  love.graphics.setColor(0, 255, 0, 255)
+  love.graphics.polygon("fill",
+    self.frontier.right.body:getWorldPoints(self.frontier.right.shape:getPoints())
   )
+  love.graphics.setColor(255, 0, 0, 255)
+  love.graphics.polygon("fill",
+    self.frontier.left.body:getWorldPoints(self.frontier.left.shape:getPoints())
+  )
+
 end
 
-function road:update(x, y, angle)
-  self.frontier.body:setAngle(angle)
-  fy = (self.roadjog) * math.sin(self.frontier.body:getAngle())
-  fx = (self.roadjog) * math.cos(self.frontier.body:getAngle())
-  self.frontier.body:setX(x + fx)
-  self.frontier.body:setY(y + fy)
+function road:update(angle, roadShift)
   self:addsegment()
+
+  mainX = self.frontier.main.body:getX()
+  mainY = self.frontier.main.body:getY()
+  
+  lastAngle = self.frontier.main.body:getAngle()
+  deltaAngle = lastAngle - angle
+  if math.abs(deltaAngle) > math.pi then
+    if deltaAngle > 0 then
+      deltaAngle = deltaAngle - 2 * math.pi
+    else
+      deltaAngle = deltaAngle + 2 * math.pi
+    end
+  end
+  
+  if math.abs(deltaAngle) > math.pi/2 then
+    --car crashes? this should never happen unless people get off the track
+  end
+  
+  if math.abs(deltaAngle) > self.rotationLimit then
+    if deltaAngle > 0 then
+      angle = lastAngle - self.rotationLimit
+    else
+      angle = lastAngle + self.rotationLimit
+    end
+  end
+  
+  self.frontier.main.body:setAngle(angle)  
+  deltaMainX = (self.roadPush) * math.cos(angle)
+  deltaMainY = (self.roadPush) * math.sin(angle)
+  self.frontier.left.body:setAngle(angle)
+  leftAngle = angle - (math.pi/2)
+  if leftAngle < -math.pi then leftAngle = leftAngle + 2 * math.pi end
+  deltaLeftX = self.frontier.main.length/3 * math.cos(leftAngle)
+  deltaLeftY = self.frontier.main.length/3 * math.sin(leftAngle)
+  self.frontier.right.body:setAngle(angle)
+  rightAngle = angle + (math.pi/2)
+  if rightAngle > math.pi then rightAngle = rightAngle - 2 * math.pi end
+  deltaRightX = self.frontier.main.length/3 * math.cos(rightAngle)
+  deltaRightY = self.frontier.main.length/3 * math.sin(rightAngle)
+  
+  if roadShift == "center" then
```

Comment:
Since we end up discarding all the calculations of `deltaSlideX`, `deltaSlideY` and `slideAngle`, should we refactor this conditional to only do those calculations in the case that `roadShift != "center"`?

Reply:
No. We have two different things going on: First is `deltaSlideX` and `deltaSlideY`, which is the sliding that happens perpendicular to the trajectory of the frontier when the player is close to the edge of the track. The second transformation is `deltaLeftX/Y` and `deltaRightX/Y` which are the transformations perpendicular to the trajectory of the frontier that happen for `self.frontier.left` and `self.frontier.right` so that they are on the edges of the frontier. These values rely on `rightAngle` and `leftAngle` and must be calculated every update.